File: 01/main.py
from typing import List


def part1(data):
    col_1 = []
    col_2 = []
    for line in data:
        a,b = line.split()
        col_1.append(int(a))
        col_2.append(int(b))
    total = 0
    for p1,p2 in zip(sorted(col_1),sorted(col_2)):
        total += abs(p1 - p2)
    print(total)


def part2(data):
    col_1 = []
    col_2 = []
    for line in data:
        a, b = line.split()
        col_1.append(int(a))
        col_2.append(int(b))
    total = 0
    counts = {}
    for item in col_1:
        total += item * counts.get(item, col_2.count(item))
    print(total)

if __name__ == '__main__':
    with open("input") as f:
        lines = f.readlines()
    part1(lines)
    part2(lines)

File: 01/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
    let path = "input";
    println!("Part 1: {}", part1(path));
    println!("Part 2: {}", part2(path));
}

struct ProblemData{
    col1: Vec<i32>,
    col2: Vec<i32>,
}

fn parse_data(path: &str) -> ProblemData {
    let file = File::open(path).unwrap();
    let reader  = BufReader::new(file);
    let mut col1 = vec![];
    let mut col2 = vec![];
    for line in reader.lines() {
        let line = line.unwrap();
        let mut split = line.split_whitespace();

        let item1 = split.next().unwrap().parse::<i32>().unwrap();
        let item2 = split.next().unwrap().parse::<i32>().unwrap();
        col1.push(item1);
        col2.push(item2);
    }
    col1.sort();
    col2.sort();
    ProblemData{col1,col2}
}

fn part1(path: &str) -> u32 {
    let data = parse_data(path);
    let mut total = 0;
    for (x,y) in data.col1.iter().zip(data.col2.iter()) {
        total += x.abs_diff(*y)
    }
    total
}

fn part2(path: &str) -> i32 {
    let data = parse_data(path);
    let mut cache = HashMap::new();
    let mut total = 0;
    for item in data.col1.iter() {
        let val = match cache.contains_key(&item){
            true => {cache.get(&item).unwrap()},
            false => {
                let val = data.col2.iter().filter(|&n| n == item).count() as i32;
                cache.insert(item, val);
                cache.get(&item).unwrap()
            }
        };
        total += item * val;
    }
    total
}